---
title: "Gout_AD"
author: "Mingzhou_Fu"
date: "7/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Part 0: Data preparation
## Load in data
```{r message=FALSE, warning=FALSE}
rm(list = ls())
library(tidyverse)
library(writexl)
library(compareGroups)
# devtools::install_github("mzfu/epiflow", upgrade_dependencies = FALSE)
library(epiflow)
library(PheWAS)
library(pROC)
library(lmtest) #coeftest
library(MatchIt)
library(sandwich) #vcovClibrary(MatchIt)
library(gtools)
library(mgcv)

work_data_path = '/Users/Mingzhou/Desktop/WQE/data/'
output_path = '/Users/Mingzhou/Desktop/WQE/Result/AD_Gout/'
phe_table_extract_date = "2021-07-09"

# Load all files (European ancestry)
load(file = paste0(work_data_path, 'prs.FULL.rda'))
load(file = paste0(work_data_path, 'safejoin_icd.rda'))
load(file = paste0(work_data_path, 'merged_demo.rda'))
clinical_var = merged_demo %>% 
  dplyr::select(UniqueSampleId, n_encounter, n_diagnosis, time_span, min_enc, max_enc)
demographic = read_csv(file = paste0(work_data_path, "BioBank_Demo.csv"))
```

## Merge DFs together
```{r message=FALSE, warning=FALSE}
pheno_merged = prs_atlas_final %>% mutate(UniqueSampleId = as.integer(IID)) %>% 
  dplyr::select(UniqueSampleId, prs_eurnorm, assigned_anc) %>% distinct() %>% 
  inner_join(safejoin_icd, by = c("UniqueSampleId" = "UniqueSampleId")) %>% distinct()
# Also add covariates: Age, Sex, PC1-5
cov_df = prs_atlas_final %>% mutate(UniqueSampleId = as.integer(IID)) %>% 
  inner_join(demographic) %>% 
  mutate(Age = round(as.numeric((as.Date(phe_table_extract_date)-as.Date(BirthDate))/365.25))) %>% 
  dplyr::select(UniqueSampleId, Age, Sex, PC1, PC2, PC3, PC4, PC5) %>% 
  left_join(clinical_var) %>% 
  distinct()
# By ancestry
prs_merged_EUR = pheno_merged %>% filter(assigned_anc == "EUR")
# Function needed
scale1 = function(x, na.rm = FALSE) (as.logical(x >= 1))
# Subset to cases >= 50, convert to logical 
pheno_df_pre = prs_merged_EUR %>% dplyr::select(-c(UniqueSampleId, prs_eurnorm, assigned_anc)) %>% 
  dplyr::select_if(function(col) sum(col, na.rm = T) >= 50) %>% mutate_all(scale1)
pheno_df = cbind(prs_merged_EUR$UniqueSampleId, pheno_df_pre) %>% rename(UniqueSampleId = V1)
prs_df = prs_merged_EUR %>% dplyr::select(UniqueSampleId, prs_eurnorm)
prs_full_EUR = pheno_df %>% inner_join(cov_df) %>% inner_join(prs_df) 
```

## Make a dataset for Gout+AD
```{r}
goutAD_EUR = prs_full_EUR %>% 
  left_join(clinical_var) %>% 
  mutate(AD = case_when(
    `290.11` == 0 ~ 0,
    `290.11` >= 1 ~ 1
  )) %>% 
  mutate(gout = case_when(
    `274.11` == 0 ~ 0,
    `274.11` >= 1 ~ 1
  )) %>% 
  mutate(overweight = case_when(
    `278` == 0 ~ 0,
    `278` >= 1 ~ 1
  )) %>% 
  mutate(hypertension = case_when(
    `401.1` == 0 ~ 0,
    `401.1` >= 1 ~ 1
  )) %>% 
  mutate(diabetes = case_when(
    `250.2` == 0 ~ 0,
    `250.2` >= 1 ~ 1
  )) %>% 
  mutate(stroke = case_when(
    `433` == 0 ~ 0,
    `433` >= 1 ~ 1
  )) %>% 
  mutate(hyperlipid = case_when(
    `272.1` == 0 ~ 0,
    `272.1` >= 1 ~ 1
  )) %>% 
  filter(Sex != "X") %>% 
  mutate(PRS_qt = quantcut(prs_eurnorm, q = 4, na.rm = T)) %>% 
  mutate(PRS_qt_low = case_when(
    prs_eurnorm <= -0.896 ~ 1,
    TRUE ~ 0
  )) %>% 
  mutate(PRS_qt_high = case_when(
    prs_eurnorm > 0.864 ~ 1,
    TRUE ~ 0
  )) %>% 
  mutate(sex = if_else(Sex == "Male", 0, 1)) %>% 
  dplyr::select(UniqueSampleId, Age, sex, PC1, PC2, PC3, PC4, PC5, prs_eurnorm, PRS_qt_low, PRS_qt_high, AD, gout, time_span,
         overweight, hypertension, diabetes, stroke, hyperlipid) 
```

# Part 1. Gout and AD PRS in EUR sample
## Prepare control and sex-specific datasets
```{r}
goutAD_EUR_male = goutAD_EUR %>% filter(sex == 0)
goutAD_EUR_female = goutAD_EUR %>% filter(sex == 1)
# Exclusion pheWAS -- exclude AD cases
goutAD_EUR_control = goutAD_EUR %>% filter(AD == 0)
```

## Logistic regression continuous
```{r message=FALSE, warning=FALSE}
# List models
lst_model = c('crude_full', 'demo_full', 'crude_control', 'demo_control',
              'crude_male', 'demo_male', 'crude_female', 'demo_female')
gene_adj_basic = 'Age + sex + PC1 + PC2 + PC3 + PC4 + PC5'
# Crude 
crude_model = as.formula(paste('gout ~ ', 'prs_eurnorm'))
crude_full = glm(crude_model, data = goutAD_EUR, family = 'binomial')
crude_control = glm(crude_model, data = goutAD_EUR_control, family = 'binomial')
crude_male = glm(crude_model, data = goutAD_EUR_male, family = 'binomial')
crude_female = glm(crude_model, data = goutAD_EUR_female, family = 'binomial')
# Adding in genetic adjusted variables
f1 = as.formula(paste('gout ~ prs_eurnorm + ', gene_adj_basic))
demo_full = glm(f1, data = goutAD_EUR, family = 'binomial')
demo_control = glm(f1, data = goutAD_EUR_control, family = 'binomial')
demo_male = glm(f1, data = goutAD_EUR_male, family = 'binomial')
demo_female = glm(f1, data = goutAD_EUR_female, family = 'binomial')
# Make OR table
gout_prs_cont = make_OR_table(lst_model, 2, 'gout_prs_')
```

## Logistic regression binary
### PRS_qt_high
```{r message=FALSE, warning=FALSE}
# List models
lst_model = c('crude_full', 'demo_full', 'crude_control', 'demo_control',
              'crude_male', 'demo_male', 'crude_female', 'demo_female')
gene_adj_basic = 'Age + sex + PC1 + PC2 + PC3 + PC4 + PC5'
# Crude 
crude_model = as.formula(paste('gout ~ ', 'PRS_qt_high'))
crude_full = glm(crude_model, data = goutAD_EUR, family = 'binomial')
crude_control = glm(crude_model, data = goutAD_EUR_control, family = 'binomial')
crude_male = glm(crude_model, data = goutAD_EUR_male, family = 'binomial')
crude_female = glm(crude_model, data = goutAD_EUR_female, family = 'binomial')
# Adding in genetic adjusted variables
f1 = as.formula(paste('gout ~ PRS_qt_high + ', gene_adj_basic))
demo_full = glm(f1, data = goutAD_EUR, family = 'binomial')
demo_control = glm(f1, data = goutAD_EUR_control, family = 'binomial')
demo_male = glm(f1, data = goutAD_EUR_male, family = 'binomial')
demo_female = glm(f1, data = goutAD_EUR_female, family = 'binomial')
# Make OR table
gout_prs_high = make_OR_table(lst_model, 2, 'gout_prs_high_')
```

### PRS_qt_low
```{r message=FALSE, warning=FALSE}
# List models
lst_model = c('crude_full', 'demo_full', 'crude_control', 'demo_control')
gene_adj_basic = 'Age + sex + PC1 + PC2 + PC3 + PC4 + PC5'
# Crude 
crude_model = as.formula(paste('gout ~ ', 'PRS_qt_low'))
crude_full = glm(crude_model, data = goutAD_EUR, family = 'binomial')
crude_control = glm(crude_model, data = goutAD_EUR_control, family = 'binomial')

# Adding in genetic adjusted variables
f1 = as.formula(paste('gout ~ PRS_qt_low + ', gene_adj_basic))
demo_full = glm(f1, data = goutAD_EUR, family = 'binomial')
demo_control = glm(f1, data = goutAD_EUR_control, family = 'binomial')

# Make OR table
gout_prs_low = make_OR_table(lst_model, 2, 'gout_prs_low_')

# Combine results
gout_prs_full = rbind(gout_prs_cont, gout_prs_high, gout_prs_low)
```

# Part 2. Gout and AD in EUR sample
## 1) Find covariates
```{r}
goutAD_EUR_cat = goutAD_EUR %>% mutate(sex_cat = if_else(sex == 0, "Male", "Female"),
                                       AD_cat = if_else(AD == 0, "Normal", "AD"),
                                       gout_cat = if_else(gout == 0, "no_gout", "gout"),
                                       overweight_cat = if_else(overweight == 0, "No", "Yes"),
                                       hypertension_cat = if_else(hypertension == 0, "No", "Yes"),
                                       diabetes_cat = if_else(diabetes == 0, "No", "Yes"),
                                       stroke_cat = if_else(stroke == 0, "No", "Yes"),
                                       hyperlipid_cat = if_else(hyperlipid == 0, "No", "Yes"))
# Build table: 1. AD status
table1_cog = descrTable(AD_cat ~ gout_cat + Age + sex_cat + time_span +
                          overweight_cat + hypertension_cat + diabetes_cat + stroke_cat + hyperlipid_cat +
                          prs_eurnorm + PC1 + PC2 + PC3 + PC4 + PC5, goutAD_EUR_cat, hide.no = "No", show.all = T)
compareGroups::export2csv(table1_cog, file = paste0(output_path, 'AD_bivariate.csv'))
# Build table: 2. Gout status
table1_gout = descrTable(gout_cat ~ AD_cat + Age + sex_cat + time_span +
                           overweight_cat + hypertension_cat + diabetes_cat + stroke_cat + hyperlipid_cat +
                          prs_eurnorm + PC1 + PC2 + PC3 + PC4 + PC5, goutAD_EUR_cat, hide.no = "No", show.all = T)
compareGroups::export2csv(table1_gout, file = paste0(output_path, 'Gout_bivariate.csv'))
```

## 2) Logistic regressions
```{r}
# Prepare non-NA datasets
goutAD_EUR_noNA = goutAD_EUR %>% dplyr::select(-overweight) %>% drop_na()
# List models
lst_model = c('crude_full', 'demo_full', 'health_full', 'prs_full')
demo_adj = 'Age + sex + PC1 + PC2 + PC3 + PC4 + PC5 + time_span'
health_adj = paste0(demo_adj, ' + hypertension + diabetes + stroke + hyperlipid')
```

### Full sample
```{r message=FALSE, warning=FALSE}
# Crude 
crude_model = as.formula(paste('AD ~ ', 'gout'))
crude_full = glm(crude_model, data = goutAD_EUR_noNA, family = 'binomial')
# Adding in genetic adjusted variables
f1 = as.formula(paste('AD ~ gout + ', demo_adj))
demo_full = glm(f1, data = goutAD_EUR_noNA, family = 'binomial')
# Adding in health adjusted variables
f2 = as.formula(paste('AD ~ gout + ', health_adj))
health_full = glm(f2, data = goutAD_EUR_noNA, family = 'binomial')
# Adjusted for PRS low
f3 = as.formula(paste('AD ~ gout + PRS_qt_low + ', health_adj))
prs_full = glm(f3, data = goutAD_EUR_noNA, family = 'binomial')
# Make OR table
AD_gout_full = make_OR_table(lst_model, 2, 'AD_gout_')
```

## Output results to file
```{r}
sheets_OR = list('gout_prs_full' = gout_prs_full,
                 'AD_gout_full' = AD_gout_full
                 )
write_xlsx(sheets_OR, path = paste0(output_path, 'AD_Gout_associations.xlsx'))
```

# Part 3. One-sample MR
```{r}
mod1 = glm(gout ~ prs_eurnorm, data = goutAD_EUR_noNA, family=binomial(link='probit'))
mod2 = glm(AD ~ mod1$fitted.values, data = goutAD_EUR_noNA, family=binomial(link='probit'))
summary(mod2)
```

```{r}
mod1 = glm(gout ~ prs_eurnorm + Age + sex + PC1 + PC2 + PC3 + PC4 + PC5 + time_span, 
           data = goutAD_EUR_noNA, family=binomial(link='probit'))
mod2 = glm(AD ~ mod1$fitted.values + Age + sex + PC1 + PC2 + PC3 + PC4 + PC5 + time_span, 
           data = goutAD_EUR_noNA, family=binomial(link='probit'))
summary(mod2)
```

```{r}
mod1 = glm(gout ~ prs_eurnorm + Age + sex + PC1 + PC2 + PC3 + PC4 + PC5 + time_span + 
             hypertension + diabetes + stroke + hyperlipid, data = goutAD_EUR_noNA, family=binomial(link='probit'))
mod2 = glm(AD ~ mod1$fitted.values + Age + sex + PC1 + PC2 + PC3 + PC4 + PC5 + time_span + 
             hypertension + diabetes + stroke + hyperlipid, data = goutAD_EUR_noNA, family=binomial(link='probit'))
summary(mod2)
```

# Part 4. Two-sample MR
```{r}
library(TwoSampleMR)
library(ieugwasr)
library(genetics.binaRies)
library(R.utils)
library(readr)
library(LDlinkR)
output_path = '/Users/Mingzhou/Desktop/WQE/Result/'
# Read in GWAS
Tin_Gout = read.table("~/Downloads/gout_Tin.txt", sep = ' ', header = T)
Tin_Gout = as.data.frame(Tin_Gout)
Kunkle_AD = read.table("~/Downloads/Kunkle_etal_Stage1_results.txt", sep = ' ', header = T)
Kunkle_AD = as.data.frame(Kunkle_AD)

# P-value setups
p_ = 1e-06
p_ = 5e-08
```


## 1. Gout on AD
### Perform MR
```{r}
# Prepare datasets to do LD pruning
exposure_rename = Tin_Gout %>% 
  rename(beta = Effect, pval = P.value, se = StdErr, chr = Chr, samplesize = n_total_sum, eaf = Freq1, 
         pos = Pos_b37, effect_allele = Allele1, other_allele = Allele2, SNP = RSID)
         
exposure_raw = exposure_rename %>% filter(pval <= p_) 
exposure = format_data(exposure_raw , type = "exposure" )
exposure$exposure = gsub("exposure", "Gout", exposure$exposure)
# Clump data
clumpedexposure = clump_data(exposure)
outcome_dat = read_outcome_data(filename = "~/Downloads/Kunkle_etal_Stage1_results.txt", snps = clumpedexposure$SNP, sep = ' ', 
                                snp_col = "MarkerName", beta_col = "Beta", se_col = "SE", effect_allele_col = "Effect_allele",
                                other_allele_col = "Non_Effect_allele", pval_col = "Pvalue", chr_col = "Chromosome", pos_col = "Position")
outcome_dat$outcome = gsub("outcome" , "AD_Kunkle", outcome_dat$outcome)
# Perform MR
dat = harmonise_data(exposure_dat = clumpedexposure, outcome_dat = outcome_dat)
mr_result = mr(dat, method_list = c("mr_ivw_fe", "mr_ivw_mre" , "mr_egger_regression" , 
                                    "mr_weighted_median", "mr_weighted_mode"))
mr_result_liberal = mr_result
mr_result_conserv = mr_result
MR_gout_on_AD = rbind(mr_result_liberal, mr_result_conserv)
```

### Assumption check
```{r}
# A) Calculate a per-SNP F statistic
dat$EAF2 = (1 - dat$eaf.exposure)
dat$MAF = pmin(dat$eaf.exposure, dat$EAF2) 
PVEfx = function(BETA, MAF, SE, N){
  pve <- (2*(BETA^2)*MAF*(1 - MAF))/
  ((2*(BETA^2)*MAF*(1 - MAF)) + ((SE^2)*2*N*MAF*(1 - MAF)))
  return(pve) 
}
n_exposure_popsize = 763813
dat$PVE = mapply(PVEfx, dat$beta.exposure, dat$MAF, dat$se.exposure, N = n_exposure_popsize)
dat$FSTAT <- ((n_exposure_popsize - 1 - 1)/1)*(dat$PVE/(1 - dat$PVE))
# B) Calculate a total instrument F statistic
((n_exposure_popsize - nrow(dat) - 1)/ nrow(dat))*(sum(dat$PVE)/(1 - sum(dat$PVE))) 
```

```{r}
# Cochran's Q
View(mr_heterogeneity(dat))
#Egger intercept
View(mr_pleiotropy_test(dat))
#I^2
Isq(dat$beta.exposure , dat$se.exposure)
#MR-PRESSO
run_mr_presso(dat)
```

## 2. AD on gout
### Perform MR
```{r}
# Prepare datasets to do LD pruning
exposure_rename = Kunkle_AD %>% 
  rename(beta = Beta, pval = Pvalue, se = SE, chr = Chromosome, 
         pos = Position, effect_allele = Effect_allele, other_allele = Non_Effect_allele, SNP = MarkerName)

exposure_raw = exposure_rename %>% filter(pval <= p_) 
exposure = format_data(exposure_raw , type = "exposure" )
exposure$exposure = gsub("exposure", "AD", exposure$exposure)
# Clump data
clumpedexposure = clump_data(exposure)
outcome_dat_raw = Tin_Gout %>% filter(RSID %in% clumpedexposure$SNP)
outcome_dat = outcome_dat_raw %>% 
  rename(beta = Effect, pval = P.value, se = StdErr, chr = Chr, 
         pos = Pos_b37, effect_allele = Allele1, other_allele = Allele2, SNP = RSID)
outcome = format_data(outcome_dat , type = "outcome" )
outcome$outcome = gsub("outcome" , "Gout", outcome$outcome)

# Perform MR
dat = harmonise_data(exposure_dat = clumpedexposure, outcome_dat = outcome)
mr_result = mr(dat, method_list = c("mr_ivw_fe", "mr_ivw_mre" , "mr_egger_regression" , 
                                    "mr_weighted_median", "mr_weighted_mode"))
mr_result_liberal = mr_result
mr_result_conserv = mr_result
MR_AD_on_gout = rbind(mr_result_liberal, mr_result_conserv)
```

### Assumption check
```{r}
# Cochran's Q
View(mr_heterogeneity(dat))
#Egger intercept
View(mr_pleiotropy_test(dat))
#I^2
Isq(dat$beta.exposure , dat$se.exposure)
#MR-PRESSO
run_mr_presso(dat)
```

## Outputs
```{r}
sheets_causal = list('MR_gout_on_AD' = MR_gout_on_AD,
                     'MR_AD_on_gout' = MR_AD_on_gout)
write_xlsx(sheets_causal, path = paste0(output_path, 'MR_result.xlsx'))
```