---
title: "Secondary associations"
author: "Mingzhou_Fu"
date: "7/21/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
library(TwoSampleMR)
library(tidyverse)
library(writexl)
# devtools::install_github("mrcieu/ieugwasr")
# devtools::install_github("explodecomputer/genetics.binaRies")
library(ieugwasr)
library(genetics.binaRies)
library(R.utils)
library(readr)
library(LDlinkR)
output_path = '/Users/Mingzhou/Desktop/WQE/Result/'
# Read in GWAS
Tin_Gout = read.table("~/Downloads/gout_Tin.txt", sep = ' ', header = T)
Tin_Gout = as.data.frame(Tin_Gout)
Kunkle_AD = read.table("~/Downloads/Kunkle_etal_Stage1_results.txt", sep = ' ', header = T)
Kunkle_AD = as.data.frame(Kunkle_AD)

# P-value setups
p_ = 1e-06
p_ = 5e-08
```

# Gout on AD
```{r}
# Prepare datasets to do LD pruning
exposure_rename = Tin_Gout %>% 
  rename(beta = Effect, pval = P.value, se = StdErr, chr = Chr, samplesize = n_total_sum, eaf = Freq1, 
         pos = Pos_b37, effect_allele = Allele1, other_allele = Allele2, SNP = RSID)
         
exposure_raw = exposure_rename %>% filter(pval <= p_) 
exposure = format_data(exposure_raw , type = "exposure" )
exposure$exposure = gsub("exposure", "Gout", exposure$exposure)
# Clump data
clumpedexposure = clump_data(exposure)
outcome_dat = read_outcome_data(filename = "~/Downloads/Kunkle_etal_Stage1_results.txt", snps = clumpedexposure$SNP, sep = ' ', 
                                snp_col = "MarkerName", beta_col = "Beta", se_col = "SE", effect_allele_col = "Effect_allele",
                                other_allele_col = "Non_Effect_allele", pval_col = "Pvalue", chr_col = "Chromosome", pos_col = "Position")
outcome_dat$outcome = gsub("outcome" , "AD_Kunkle", outcome_dat$outcome)
# Perform MR
dat = harmonise_data(exposure_dat = clumpedexposure, outcome_dat = outcome_dat)
mr_result = mr(dat, method_list = c("mr_ivw_fe", "mr_ivw_mre" , "mr_egger_regression" , 
                                    "mr_weighted_median", "mr_weighted_mode"))
mr_result_liberal = mr_result
mr_result_conserv = mr_result
MR_gout_on_AD = rbind(mr_result_liberal, mr_result_conserv)
```

```{r}
# A) Calculate a per-SNP F statistic
dat$EAF2 = (1 - dat$eaf.exposure)
dat$MAF = pmin(dat$eaf.exposure, dat$EAF2) 
PVEfx = function(BETA, MAF, SE, N){
  pve <- (2*(BETA^2)*MAF*(1 - MAF))/
  ((2*(BETA^2)*MAF*(1 - MAF)) + ((SE^2)*2*N*MAF*(1 - MAF)))
  return(pve) 
}
n_exposure_popsize = 763813
dat$PVE = mapply(PVEfx, dat$beta.exposure, dat$MAF, dat$se.exposure, N = n_exposure_popsize)
dat$FSTAT <- ((n_exposure_popsize - 1 - 1)/1)*(dat$PVE/(1 - dat$PVE))
# B) Calculate a total instrument F statistic
((n_exposure_popsize - nrow(dat) - 1)/ nrow(dat))*(sum(dat$PVE)/(1 - sum(dat$PVE))) 
```

```{r}
# Cochran's Q
View(mr_heterogeneity(dat))
#Egger intercept
View(mr_pleiotropy_test(dat))
#I^2
Isq(dat$beta.exposure , dat$se.exposure)
#MR-PRESSO
run_mr_presso(dat)
```


# AD on gout
```{r}
# Prepare datasets to do LD pruning
exposure_rename = Kunkle_AD %>% 
  rename(beta = Beta, pval = Pvalue, se = SE, chr = Chromosome, 
         pos = Position, effect_allele = Effect_allele, other_allele = Non_Effect_allele, SNP = MarkerName)

exposure_raw = exposure_rename %>% filter(pval <= p_) 
exposure = format_data(exposure_raw , type = "exposure" )
exposure$exposure = gsub("exposure", "AD", exposure$exposure)
# Clump data
clumpedexposure = clump_data(exposure)
outcome_dat_raw = Tin_Gout %>% filter(RSID %in% clumpedexposure$SNP)
outcome_dat = outcome_dat_raw %>% 
  rename(beta = Effect, pval = P.value, se = StdErr, chr = Chr, 
         pos = Pos_b37, effect_allele = Allele1, other_allele = Allele2, SNP = RSID)
outcome = format_data(outcome_dat , type = "outcome" )
outcome$outcome = gsub("outcome" , "Gout", outcome$outcome)

# Perform MR
dat = harmonise_data(exposure_dat = clumpedexposure, outcome_dat = outcome)
mr_result = mr(dat, method_list = c("mr_ivw_fe", "mr_ivw_mre" , "mr_egger_regression" , 
                                    "mr_weighted_median", "mr_weighted_mode"))
mr_result_liberal = mr_result
mr_result_conserv = mr_result
MR_AD_on_gout = rbind(mr_result_liberal, mr_result_conserv)
```

```{r}
# Cochran's Q
View(mr_heterogeneity(dat))
#Egger intercept
View(mr_pleiotropy_test(dat))
#I^2
Isq(dat$beta.exposure , dat$se.exposure)
#MR-PRESSO
run_mr_presso(dat)
```


## Outputs
```{r}
sheets_causal = list('MR_gout_on_AD' = MR_gout_on_AD,
                     'MR_AD_on_gout' = MR_AD_on_gout)
write_xlsx(sheets_causal, path = paste0(output_path, 'MR_result.xlsx'))
```





## Gout on AD
```{r}
# Diagnoses - secondary ICD10: M10.99 Gout, unspecified (Site unspecified) -- UKB (2018)
p1 = c(0.1, 0.3, 1)
p2 = c(0.001, 0.01, 0.1)
r2 = c(0.001, 0.01, 0.2)


MR_GUGC_grid = data.frame(p1_par = double(), p2_par = double(), r2_par = double(),
                          outcome = character(), exposure = character(), method = character(), 
                          nsnp = double(), b = double(), se = double(), pval = double(), 
                          OR = double(), lowCI = double(), highCI = double(), stringsAsFactors = FALSE)
for (i in 1:length(p1)) {
  p1_ = p1[i]
  for (m in 1:length(p2)) {
    p2_ = p2[m]
    for (n in 1:length(r2)) {
      r2_ = r2[n]
      print(paste0("p1: ", p1_, " , p2: ", p2_, " , r2: ", r2_))
      if (p1_ == 1) {
        clump_ = FALSE
      } else {
        clump_ = TRUE
      }
      exposure_dat_GUGC = extract_instruments("ukb-b-12765", p1 = p1_, clump = clump_, 
                                              p2 = p2_, r2 = r2_, kb = 10000) 
      outcome_dat_GUGC = extract_outcome_data(snps = exposure_dat_GUGC$SNP, outcomes = "ieu-b-2")
      dat_GUGC = harmonise_data(exposure_dat_GUGC, outcome_dat_GUGC)
      res_GUGC = mr(dat_GUGC)
      res_GUGC_add = res_GUGC %>% mutate(OR = exp(b)) %>% 
        mutate(lowCI = exp(b - 1.96*se),
               highCI = exp(b + 1.96*se),
               p1_par = p1_,
               p2_par = p2_,
               r2_par = r2_) %>% 
        select(p1_par, p2_par, r2_par, outcome, exposure, method, nsnp, b, se, pval, OR, lowCI, highCI)
      MR_GUGC_grid = rbind(MR_GUGC_grid, res_GUGC_add)
      n = n + 1
    }
    m = m + 1
  }
  i = i + 1
}




exposure_dat_UKB = extract_instruments("ukb-b-12765", p1 = 0.05, clump = TRUE, p2 = 0.01, r2 = 0.001, kb = 10000) 
# Non-cancer illness code, self-reported: gout
exposure_dat_UKB_self = extract_instruments("ukb-b-13251", p1 = 0.05, clump = TRUE, p2 = 0.01, r2 = 0.001, kb = 10000) 
# Kottgen et al. 2013
exposure_dat_GUGC = extract_instruments("ieu-a-1054", p1 = 0.05, clump = TRUE, p2 = 0.01, r2 = 0.001, kb = 10000) 

# Get effects of instruments on outcome
outcome_dat_UKB = extract_outcome_data(snps = exposure_dat_UKB$SNP, outcomes = "ieu-b-2") # Kunkle,2019
outcome_dat_UKB_self = extract_outcome_data(snps = exposure_dat_UKB_self$SNP, outcomes = "ieu-b-2") # Kunkle,2019
outcome_dat_GUGC = extract_outcome_data(snps = exposure_dat_GUGC$SNP, outcomes = "ieu-b-2") # Kunkle,2019
```

```{r}
# Harmonise the exposure and outcome data
dat_UKB = harmonise_data(exposure_dat_UKB, outcome_dat_UKB)
dat_self = harmonise_data(exposure_dat_UKB_self, outcome_dat_UKB_self)
dat_GUGC = harmonise_data(exposure_dat_GUGC, outcome_dat_GUGC)
# Perform MR
res_UKB = mr(dat_UKB)
res_UKB_add = res_UKB %>% mutate(OR = exp(b)) %>% 
  mutate(lowCI = exp(b - 1.96*se)) %>% 
  mutate(highCI = exp(b + 1.96*se)) %>% 
  select(outcome, exposure, method, nsnp, b, se, pval, OR, lowCI, highCI)
res_self = mr(dat_self)
res_self_add = res_self %>% mutate(OR = exp(b)) %>% 
  mutate(lowCI = exp(b - 1.96*se)) %>% 
  mutate(highCI = exp(b + 1.96*se)) %>% 
  select(outcome, exposure, method, nsnp, b, se, pval, OR, lowCI, highCI)
res_GUGC = mr(dat_GUGC)
res_GUGC_add = res_GUGC %>% mutate(OR = exp(b)) %>% 
  mutate(lowCI = exp(b - 1.96*se)) %>% 
  mutate(highCI = exp(b + 1.96*se)) %>% 
  select(outcome, exposure, method, nsnp, b, se, pval, OR, lowCI, highCI)


```

# AD on gout
```{r}
# Diagnoses - secondary ICD10: M10.99 Gout, unspecified (Site unspecified) -- UKB (2018)
exposure_dat_UKB = extract_instruments("ukb-b-12765 ieu-b-2", p1 = 0.05, clump = TRUE, p2 = 0.01, r2 = 0.001, kb = 10000) 
# Get effects of instruments on outcome
outcome_dat_UKB = extract_outcome_data(snps = exposure_dat_UKB$SNP, outcomes = "") # Kunkle,2019
```




